modules/kvtagger/kvtagdb.c

```diff
@@ -0,0 +1,264 @@
+ * COPYING for details.
+ *
+ */
+
+#include "kvtagdb.h"
+#include "atomic.h"
+#include <string.h>
+#include <sys/types.h>
+
+struct _KVTagDB
+{
+  GAtomicCounter ref_cnt;
+  GArray *data;
+  GHashTable *index;
+  gboolean is_data_indexed;
+};
+
+typedef struct _element_range
+{
+  gsize offset;
+  gsize length;
+} element_range;
+
+static gint
+_tagrecord_cmp(gconstpointer k1, gconstpointer k2)
+{
+  TagRecord *r1 = (TagRecord *)k1;
+  TagRecord *r2 = (TagRecord *)k2;
+
+  return strcmp(r1->selector, r2->selector);
+}
+
+void
+kvtagdb_index(KVTagDB *self)
+{
+  if (self->data->len > 0)
+    {
+      g_array_sort(self->data, _tagrecord_cmp);
+      gsize range_start = 0;
+      TagRecord range_start_tag = g_array_index(self->data, TagRecord, 0);
+
+      for (gsize i = 1; i < self->data->len; ++i)
+        {
+          TagRecord current_record = g_array_index(self->data, TagRecord, i);
+
+          if (_tagrecord_cmp(&range_start_tag, &current_record))
+            {
+              element_range *current_range = g_new(element_range, 1);
+              current_range->offset = range_start;
+              current_range->length = i - range_start;
+
+              g_hash_table_insert(self->index, range_start_tag.selector, current_range);
+
+              range_start_tag = current_record;
+              range_start = i;
+            }
+        }
+
+      {
+        element_range *last_range = g_new(element_range, 1);
+        last_range->offset = range_start;
+        last_range->length = self->data->len - range_start;
+        g_hash_table_insert(self->index, range_start_tag.selector, last_range);
+      }
+      self->is_data_indexed = TRUE;
+    }
+}
+
+static void
+_ensure_indexed_db(KVTagDB *self)
+{
+  if (!self->is_data_indexed)
+    kvtagdb_index(self);  
+}
+
+static void
+_tag_record_free(gpointer p)
+{
+  TagRecord *rec = (TagRecord *)p;
+  tag_record_clean(rec);
+}
+
+static void
+_new(KVTagDB *self)
+{
+  self->data = g_array_new(FALSE, FALSE, sizeof(TagRecord));
+  g_array_set_clear_func(self->data, _tag_record_free);
+  self->index = g_hash_table_new_full(g_str_hash, g_str_equal, NULL, g_free);
+  self->is_data_indexed = FALSE;
+  g_atomic_counter_set(&self->ref_cnt, 1);
+}
+
+static void
+_free(KVTagDB *self)
+{
+  if (self->index)
+    g_hash_table_unref(self->index);
+  if (self->data)
+    g_array_unref(self->data);
+}
+
+KVTagDB*
+kvtagdb_new()
+{
+  KVTagDB *self = g_new0(KVTagDB, 1);
+    
+  _new(self);  
+
+  return self;
+}
+
+KVTagDB*
+kvtagdb_ref(KVTagDB *self)
+{
+  g_assert(!self || g_atomic_counter_get(&self->ref_cnt) > 0);
+  g_atomic_counter_inc(&self->ref_cnt);
+
+  return self;
+}
+
+void
+kvtagdb_unref(KVTagDB *self)
+{
+  g_assert(!self || g_atomic_counter_get(&self->ref_cnt));
+  if (g_atomic_counter_dec_and_test(&self->ref_cnt))
+    {
+      kvtagdb_free(self);
+    }
+}
+
+static element_range*
+_get_range_of_tags(KVTagDB *const self, const gchar *const selector)
```

Comment:
I still agree that we could and should add `const` where ever possible, however because you never use `*const self` anywhere else, it seems a bit confusing here. The reader could get puzzled whether the `const` was misplaced or not.